WIP - Added main CLB outbound processing flow and configuration

georgeedward2000 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

georgeedward2000 committed:
WIP - Added main CLB outbound processing flow and configuration

georgeedward2000 committed:
WIP - Added main CLB outbound processing flow and configuration

georgeedward2000 committed:
Merge branch 'clb-combined-prs-outbound' of https://github.com/georgeedward2000/cloud-provider-azure into clb-combined-prs-outbound

georgeedward2000 committed:
remove file

georgeedward2000 committed:
start background processing worker goroutine

georgeedward2000 committed:
generated azclient.natgatewayclient files

david-kow commented on pkg/provider/azure.go:
```diff
@@ -799,6 +810,11 @@ func (az *Cloud) SetInformers(informerFactory informers.SharedInformerFactory) {
 
 	az.setUpEndpointSlicesInformer(informerFactory)
+
+	if az.ServiceGatewayEnabled || az.IsLBBackendPoolTypePodIPAndUseStandardV2LoadBalancer() {
```
What happens if only one is true? Should we validate this somewhere early and rely on a single check so that we don't run into unexpected init issues (eg. sgw is enabled, but pool type is not podip)

georgeedward2000 replied:
Let's discuss about it.

david-kow commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+					prevEgressGatewayName string
+					currEgressGatewayName string
+				)
+				if oldPod.Labels == nil || oldPod.Labels[consts.PodLabelServiceEgressGateway] == "" {
```
alternatively:

```go
				if oldPod.Labels != nil {
					prevEgressGatewayName = oldPod.Labels[consts.PodLabelServiceEgressGateway]
				}
```

david-kow commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+				} else {
+					prevEgressGatewayName = oldPod.Labels[consts.PodLabelServiceEgressGateway]
+				}
+				if newPod.Labels == nil || newPod.Labels[consts.PodLabelServiceEgressGateway] == "" {
```
same as above

david-kow commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+				}
+
+				staticGatewayConfigurationName := pod.Labels[consts.PodLabelServiceEgressGateway]
+				counter, ok := az.localServiceNameToNRPServiceMap.Load(staticGatewayConfigurationName)
```
Is there a reason why `podEgressQueue` and `localServiceNameToNRPServiceMap` are not encapsulated within diff tracker? It seems their behavior is tied to it and the only usages end up with interacting with difftracker anyway. Wouldn't moving it inside not make the code here cleaner? So basically, the interaction could be only something like:

```go
staticGatewayConfigurationName := pod.Labels[consts.PodLabelServiceEgressGateway]
az.diffTracker.UpdateK8sPod(
  difftracker.UpdatePodInputType{
	  PodOperation:           difftracker.ADD,
	  PublicOutboundIdentity: staticGatewayConfigurationName,
	  Location:               pod.Status.HostIP,
	  Address:                pod.Status.PodIP,
  },
  )
```

Updating counts and creating pod egress queue event could happen inside diff tracker.

georgeedward2000 replied:
Let's discuss about this one.

david-kow commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+					default:
+						// channel is full, do nothing
+						klog.V(2).Info("az.locationAndNRPServiceBatchUpdater.channelUpdateTrigger is full. Batch update is already triggered.")
+					}
```
I'd either make it auto trigger after updating pod or at least create a helper func to do it. The caller shouldn't need to know right semantics and channel handling to trigger it.

kartickmsft commented on pkg/provider/azure.go:
```diff
@@ -799,6 +810,11 @@ func (az *Cloud) SetInformers(informerFactory informers.SharedInformerFactory) {
 	az.setUpEndpointSlicesInformer(informerFactory)
+
+	if az.ServiceGatewayEnabled || az.IsLBBackendPoolTypePodIPAndUseStandardV2LoadBalancer() {
+		az.podLister = informerFactory.Core().V1().Pods().Lister()
```
This should be moved inside the setupPodInformerForEgress and use the podInformerFactory in that function.

kartickmsft commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+								// channel is full, do nothing
+								klog.V(2).Info("az.locationAndNRPServiceBatchUpdater.channelUpdateTrigger is full. Batch update is already triggered.")
+							}
+							az.localServiceNameToNRPServiceMap.Store(prevEgressGatewayName, counter.(int)-1) // TBD (enechitoaia): Not present in LLD. Check if this is correct
```
This is correct.

kartickmsft commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+
+		switch event.EventType {
+		case "Add":
+		case "Update":
```
I think we are enqueuing only Add/Delete Event. So, can remove handling of Update to avoid confusion.

kartickmsft commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+						klog.V(2).Info("az.locationAndNRPServiceBatchUpdater.channelUpdateTrigger is full. Batch update is already triggered.")
+					}
+				} else {
+					updater.az.diffTracker.UpdateK8sPod(difftracker.UpdatePodInputType{
```
We are enqueuing Delete event only if count is 1. The else block seems redundant and can be removed.

kartickmsft commented on pkg/provider/azure_location_and_nrp_service_batch_updater.go:
```diff
@@ -293,3 +300,377 @@ func (updater *locationAndNRPServiceBatchUpdater) removeOperation(name string) {
+				}
+			}
+		case "Delete":
+			counter, ok := updater.az.localServiceNameToNRPServiceMap.Load(service)
```
I think we need not check if service is present in map as Delete event is sent when counter is 1. IMO, we could also do the REMOVE operation in the informer delete handler. Here, just invoke NRP calls to delete/disassociate NRP resources and remove the service itself from localServiceNameToNRPServiceMap after NRP calls succeed.

georgeedward2000 committed:
0. Addressed comments
1. Moved logic from cloud into difftracker
2. Created helper methods
3. Cleaned code
4. Update tests

georgeedward2000 committed:
nil pointer exception issue solution

georgeedward2000 committed:
update instantiation order flow for CLB
